[호두] 뷰 챌린지 미션 2, 3단계 제출합니다.

Junyoung-WON opened the pull request:
안녕하세요, 차람! 호두두입니다! 🙇‍♂️
차람과의 재미있는 리뷰가 끝나가다니 아쉬워요...(사실 많이 주고받지는 못했지만..)

우선 필수 요구사항만 만족하였습니다...!
몇 가지 구현하지 못해 아쉬운 점들은,
- **Configuration Change** 에 대응하지 못해, 화면 회전 또는 다크모드 변환 시에 _기존 그림들이 모두 사라진다_ 는 것과
- 색상 및 그리기 모드 선택 시 **UI 변화** 를 구현하지 못한 것
- 그리고 그리기 모드에 따라 펜 굵기, 색상 선택 등 보여주지 않아도 되는 도구를 구분하지 않아 불편하다는 것입니다.

아래는 화면 크기 대응, 다크모드 대응 시연 이미지입니다!

## 휴대폰 화면

### 기본 화면(세로 모드)
  ![phone-portrait](https://github.com/user-attachments/assets/0f2028ef-f5bc-472f-8c8a-9c6d2cc5b84b)


### 가로 모드
  ![phone-land](https://github.com/user-attachments/assets/fb4e423c-ffc5-4520-89bd-c828cef182fc)


### 다크 모드
  ![phone-dark](https://github.com/user-attachments/assets/7fe8393a-4127-49a8-8620-5526949be666)


## 태블릿 화면
### 기본 화면(가로 모드)
  ![tablet-land](https://github.com/user-attachments/assets/47d8a7da-4245-445c-b8b8-695f18148f84)


### 세로 모드
  ![tablet-portrait](https://github.com/user-attachments/assets/7d4b7b9e-a546-4bff-8f65-14d963af33c3)


### 다크 모드
  ![tablet-land-dark](https://github.com/user-attachments/assets/4f5d95f8-35f5-41f9-a6c8-24809395b3c1)

Junyoung-WON committed:
refactor: (1단계) 눌렀을 때(ACTION_DOWN)에 그림이 그려지기 시작하도록 개선

Junyoung-WON committed:
docs: (2단계) 기능 요구 사항 작성

Junyoung-WON committed:
ui: (2단계) 색상 선택기 중앙 정렬

Junyoung-WON committed:
ui: (2단계) 테마 색상 변경

Junyoung-WON committed:
ui: (2단계) 선 굵기 컨트롤러에 좌우 마진 추가

Junyoung-WON committed:
ui: (2단계) 그리기 모드 아이콘 drawable 추가

Junyoung-WON committed:
ui: (2단계) 그리기 모드 아이템 레이아웃 추가

Junyoung-WON committed:
ui: (2단계) 그리기 모드 선택 RecyclerView 추가

Junyoung-WON committed:
ui: (2단계) 지우개 아이콘 drawable 변경

Junyoung-WON committed:
feat: (2단계) 그리기 모드 모델 추가

- 그리기 모드에 들어갈 문자열 리소스 작성

Junyoung-WON committed:
feat: (2단계) 그리기 모드 선택에 대한 클릭 핸들러 추가

Junyoung-WON committed:
feat: (2단계) 그리기 모드 RecyclerView 어댑터, 뷰홀더 작성

Junyoung-WON committed:
feat: (2단계) 그리기 모드에 따라 그림을 그리는 방법 변경하는 기능 구현

- drawingMode 에 따라 그리기 방식 분기 처리
- setLayerType : 하드웨어 가속 설정으로 지우개 구현 및 draw 연산 속도 개선 시도
- ACTION_DOWN 시 터치한 위치를 저장하여, 해당 위치를 기준으로 도형 그리기
- 원, 사각형을 그리는 로직 구현
- ACTION_DOWN 시에 그림을 추가 및 onDraw 의 canvas draw 로직 변경
- 그리기 모드를 변경하는 메서드 작성
- 메서드 분리 리팩터링

Junyoung-WON committed:
feat: (2단계) DrawMode와 DrawingModeHandler 데이터 바인딩 및 바인딩어댑터 설정

- BindingAdapters.kt: drawableRes로부터 icon을 설정하기 위해 바인딩 어댑터 생성
- build.gradle.kts: 어댑터를 사용하기 위한 kapt 설정

Junyoung-WON committed:
feat: (2단계) DrawingModeAdapter 설정 및 DefaultColor 생성

- DrawingModeHandler 구현하여 그리기 모드 변경 로직 작성
- DefaultColor 를 정의하여 기본 색상 리스트 반환

Junyoung-WON committed:
docs: (2단계) 기능 요구 사항 만족 여부 확인

- 선택 요구 사항을 제외하고 요구 사항 만족

Junyoung-WON committed:
docs: (3단계) 기능 요구 사항 작성

Junyoung-WON committed:
ui: (3단계) 색상 변경

- 색들이 너무 쨍하네요

Junyoung-WON committed:
ui: (3단계) 회색 색상 일부 수정

- 가시성을 높이기 위해 밝은 회색과 회색을 분리

Junyoung-WON committed:
ui: (3단계) 도구 패널 배경 생성

Junyoung-WON committed:
feat: (3단계) 불필요한 데이터 바인딩 제거 및 View 구성 변경

- ColorPickerAdapter 데이터바인딩 제거
- include 로 활용하던 색상 선택 RecyclerView 를 통일
- 도구 패널을 ConstraintLayout으로 묶음
  - 도구 패널의 배경 지정
  - maxWidth를 600dp로 지정
- margin, width 설정 등 조정

Junyoung-WON committed:
feat: (3단계) 태블릿 화면의 가로 모드에서는 도구 패널이 좌상단으로 이동

Junyoung-WON committed:
ui: (3단계) dimens 설정으로 가로 모드, 태블릿 화면의 가로모드 대응

Junyoung-WON committed:
ui: (3단계) 다크모드 대응 및 스타일 정의를 위한 colors 리소스 정의

Junyoung-WON committed:
ui: (3단계) 다크모드 colors 리소스 정의 및 다크모드가 적용된 스타일 정의

Junyoung-WON committed:
ui: (3단계) 테마 다크모드 대응 및 툴바 다크모드 대응

Junyoung-WON committed:
ui: (3단계) 도구 패널의 배경 색상 다크모드 대응

Junyoung-WON committed:
ui: (3단계) 스타일 및 색상 리소스 적용하여 다크모드 대응

Junyoung-WON committed:
ui: (3단계) 사용하지 않는 레이아웃과 테마 제거

Junyoung-WON committed:
docs: (3단계) 기능 요구 사항 만족 여부 확인

aprilgom commented on app/src/main/java/woowacourse/paint/model/DrawingMode.kt:
```diff
@@ -0,0 +1,54 @@
+    val xfermode: Xfermode?,
+) {
+    PEN(
+        R.drawable.icon_pencil,
```
아이콘까지 써서 이쁘게 만들어주셨네요

Junyoung-WON replied:
더 예쁘게 만들지 못해 아쉽습니다...😂

aprilgom commented on app/src/main/java/woowacourse/paint/ui/DrawingBoardView.kt:
```diff
@@ -9,53 +9,121 @@ import android.util.AttributeSet
     init {
         isFocusable = true
         isFocusableInTouchMode = true
+        setLayerType(LAYER_TYPE_HARDWARE, null)
```
얘는 뭐하는애인가요
하드웨어 가속인가요?

Junyoung-WON replied:
네, 그렇습니다! 그림판 View에 대해서 **하드웨어(GPU) 가속 설정**을 하는 코드입니다.

`setLayerType` 메서드를 이용하면, 해당 뷰가 그려질 때에 밑바탕이 될 `Layer`의 타입을 지정할 수 있습니다.
첫번째 파라미터에 들어갈 값은 `Layer` 의 타입이고, 두번째 파라미터에 들어가는 값은 `Paint` 입니다.

`Paint`는 nullable이고, `Paint`를 넘겨주었을 때 `Layer`에 적용되는 속성은 아래 3가지라고 합니다.
- Translucency (alpha) : 투명도
- Blending mode : xfermode(픽셀의 연산 및 혼합 방식)
- Color filter

`Layer`의 타입으로는 `LAYER_TYPE_NONE`, `LAYER_TYPE_SOFTWARE`, `LAYER_TYPE_HARDWARE` 이렇게 3가지가 있습니다.

### `LAYER_TYPE_NONE`
Layer를 지정하지 않습니다.
Layer가 없기 때문에, 이 경우 Layer에 대한 Paint 설정은 무시됩니다!

### `LAYER_TYPE_SOFTWARE`
Layer를 소프트웨어(CPU)를 기반으로 렌더링하도록 설정합니다.
  - 하드웨어 가속 설정을 하지 않은 경우, 해당 뷰와 모든 자식 뷰들에게 draw 관련 설정을 할 때 사용됩니다.
    > When the application is not using hardware acceleration, a software layer is useful to apply a specific color filter and/ or blending mode and/ or translucency to a view and all its children
  - 하드웨어 가속 설정을 사용하는 경우, 하드웨어 가속 파이프라인으로 지원되지 않는 자원들을 그리거나, 복잡한 뷰 트리를 캐싱하고 그리기 연산의 복잡성을 줄이는 데에 사용됩니다.
    > When the application is using hardware acceleration, a software layer is useful to render drawing primitives not supported by the hardware accelerated pipeline. It can also be used to cache a complex view tree into a texture and reduce the complexity of drawing operations. For instance, when animating a complex view tree with a translation, a software layer can be used to render the view tree only once.

즉, 뷰 트리 구조를 새로 생성하고 캐싱하여 재사용할 때 유용하게 사용한다고 합니다!

### `LAYER_TYPE_HARDWARE`
Layer를 하드웨어(GPU)를 기반으로 렌더링하도록 설정합니다.
  - 이 Layer 타입도 `LAYER_TYPE_SOFTWARE` 과 마찬가지로, **자신과 내부 뷰들의 계층 구조의 draw 관련 설정을 적용**하고, **복잡한 뷰 트리를 생성 및 캐싱**하는 데 사용됩니다.
    > A hardware layer is useful to apply a specific color filter and/ or blending mode and/ or translucency to a view and all its children.
A hardware layer can be used to cache a complex view tree into a texture and reduce the complexity of drawing operations. For instance, when animating a complex view tree with a translation, a hardware layer can be used to render the view tree only once.
  - 그리고 **픽셀의 Blending 연산과 같은 복잡한 연산을 지원**해줍니다. 그래서 **[이번 미션에서 사용한 지우개의 동작](https://github.com/woowacourse/android-paint/pull/100#discussion_r1789734241)** 을 이 설정으로 구현할 수 있었던 것이죠!
  - 또한 뷰에 회전이 적용될 때 렌더링 퀄리티를 높일 수 있다고 합니다.
    > A hardware layer can also be used to increase the rendering quality when rotation transformations are applied on a view. It can also be used to prevent potential clipping issues when applying 3D transforms on a view.

### 소프트웨어 기반 렌더링 모델 (`LAYER_TYPE_SOFTWARE`)
**소프트웨어 기반 렌더링 모델**(`LAYER_TYPE_SOFTWARE` 적용)은 아래 단계를 거쳐 View를 그립니다.
1. Invalidate the hierarchy : 뷰 계층구조를 무효화(`invalidate()`) 함
2. Draw the hierarchy : 뷰 계층구조를 그림(`onDraw()`)

위와 같은 동작 방식으로 인해서, **소프트웨어 기반 렌더링 모델**은 2가지 문제점이 있습니다.
- 특정 뷰에서 그리기 연산이 필요하여 `invalidate()`가 호출되면, **해당 계층구조의 모든 View에게 `invalidate()`가 전파**되어, 그리기 연산이 필요하지 않은 **다른 View들도 새로 그려지게 됩니다.** 그래서 불필요한 그리기 연산이 나타나게 됩니다!
- 또한 이러한 문제점으로 인해, 코드 상에서 `invalidate()`를 실수로 넣지 않아서 사실 상 버그가 일어나야하는 View도, 다른 View의 `invalidate()` 호출로 인해 새로 그려질 수 있기 때문에, **원래라면 나타나야 할 UI 오류가 숨겨질 수 있습니다.** 다른 View에서의 `invalidate()` 호출이 이루어지는 조건이 다르기 때문에, 어떤 상황에서 버그가 나타날지 불확실해지고, 그러면 나중에 이 버그를 확인하게 됐을 때 원인을 찾기 힘들 수 있겠죠!

그러하여 원하는 View에 대해서만 변경사항(그리기 연산)을 업데이트하기 위해 등장한 새로운 그리기 모델이 **하드웨어 기반 렌더링 모델**(`LAYER_TYPE_HARDWARE` 적용) 입니다!

### 하드웨어 기반 렌더링 모델(`LAYER_TYPE_HARDWARE`)

**하드웨어 기반 렌더링 모델**은 아래 3단계에 따라 View를 그립니다.
1. Invalidate the hierarchy : 뷰 계층구조를 무효화(`invalidate()`) 함
2. Record and update display lists : `display lists` 에 저장(캐싱) 및 업데이트
3. Draw the display lists : `display lists` 를 그림

여기서 `display lists`란 뷰를 렌더링하는 명령어를 미리 저장해놓은 캐시입니다.
안드로이드 시스템은 뷰를 그릴 때 바로 그리는 것이 아니라, 이 display list를 확인하고서 뷰를 그려냅니다.

**안드로이드 시스템이 View를 그리는 동작 방식**입니다 :
- 안드로이드 시스템은 `display lists` 에 View를 그리는 코드(명령어)를 `display lists` 라는 곳에 저장하여 캐싱하고, 
- `invalidate()`가 호출된 View에 대해서만 `display lists`에 `dirty` 마크를 합니다. 
- 시스템이 `dirty` 마크가 표시된 View만 확인해서 View를 그리는 명령어를 업데이트하고,
- 업데이트된 `display lists`를 바탕으로 그리기 연산을 수행합니다.
- `dirty` 마크가 없는 다른 View들은 단순히 `display list`에 있던 그리기 명령어를 재사용합니다!

소프트웨어 렌더링의 경우 위의 동작 방식을 따르지 않고, 계층구조 내 모든 View들의 display lists를 새로 계산하여 만들어내므로, 반복적인 UI 변경이 자주 발생하는 뷰의 경우에는 최적화가 어렵고 성능을 저하시킬 수 있습니다.
하드웨어 가속 설정이 된 경우 위 방식으로 최적화를 수행하기 때문에, 하드웨어 가속 설정으로 매번 모든 뷰 계층이 새로 그려지는 불필요한 연산을 막을 수 있습니다!

아래 레퍼런스를 참고하여 공부해보았습니다!
하드웨어 가속 설정 방법, 버전에 따른 하드웨어 분기 처리 등등 더 자세한 정보들이 있으니
만약 궁금하시다면 아래 공식 문서를 참고해주세요!
- [공식 문서 - Hardware Acceleration](https://developer.android.com/develop/ui/views/graphics/hardware-accel)

aprilgom replied:
공식문서보다 훨씬 나은데요???

aprilgom commented on app/src/main/java/woowacourse/paint/model/DrawingMode.kt:
```diff
@@ -0,0 +1,54 @@
+        R.string.drawing_mode_erase,
+        Style.STROKE,
+        PorterDuffXfermode(
+            PorterDuff.Mode.CLEAR
```
![img1 daumcdn](https://github.com/user-attachments/assets/8b3561a9-44ba-4c21-adb9-a82c5722d734)
PorterDuff의 16가지 mode를 가져왔습니다.

aprilgom replied:
Clear 말고 다른 모드도 써볼 수 있을 것 같아요 (Clear를 쓰지 말라는 코멘트는 아닙니다)

aprilgom replied:
지우개를 구현하면서 들었던 개인적인 고민입니다.
조금 쓸데없는 고민이니 깊게 생각 안하셔도 됩니다.
포토샵같은 툴을 쓸 때, 지우개로 지우면 부드럽게 지워지는걸 볼 수 있는데요,
이런 지우개는 PorterDuff를 이용해서 구현할 수 있을까요??

aprilgom replied:
![지우개애앵](https://github.com/user-attachments/assets/8dbd8286-0ae2-4101-8e03-326be4081ffa)

Junyoung-WON replied:
좋은 레퍼런스 너무 감사드려요!!!
다른 연산 방식이 어떤게 있는지 까지는 뜯어보지 않았는데, 차람의 딥다이브는 어디까지인가요...😮

확인해보니, `CLEAR` 말고도 `DST_OUT` 을 사용해서 구현할 수 있네요!
이 외에도 `XOR`, `SRC_OUT` 을 활용해보았는데요. 두 연산 방식의 경우, 처음 지우개를 그렸을 때는 정상적으로 지워지는 듯 했지만,
지우개 Path 위에 한번 더 겹쳐 그리는 경우에는 현재 설정되어있는 색상이 그려졌습니다.
이미 한번 지운 부분 위에 새로 그리는 경우, 기존 Destination에는 투명화 처리된 픽셀이 존재하기 때문에, Source 픽셀(색을 가지고 있음)이 해당 영역 위에 그려지게 된 것입니다.

Junyoung-WON replied:
> 포토샵같은 툴을 쓸 때, 지우개로 지우면 부드럽게 지워지는걸 볼 수 있는데요,
이런 지우개는 `PorterDuff`를 이용해서 구현할 수 있을까요??

네, 가능합니다! 마찬가지로 `DST_OUT` 과 `CLEAR` 모드로 구현할 수 있었어요.
하지만 `PorterDuff` 만을 이용해서는 구현이 어려웠습니다!
주변이 부드럽게 지워지는 효과는 블러 처리를 해야 하는데, 그러기 위해서는 `Paint`의 `maskFilter` 를 `BlurMaskFilter` 로 설정해주어야 했습니다.
- 블러 효과 설정
  ![image](https://github.com/user-attachments/assets/5d8a6a72-0e4c-44e4-90b6-38ce5b2a1a50)
- 결과
  ![image](https://github.com/user-attachments/assets/053090ca-2502-4e04-9e61-c7ac3d64b89e)

aprilgom replied:
paint에 블러를 주는것 만으로 쉽게 구현이 되는군요!! 👍👍👍

aprilgom commented on app/src/main/java/woowacourse/paint/ui/DrawingBoardView.kt:
```diff
@@ -9,53 +9,121 @@ import android.util.AttributeSet
 class DrawingBoardView constructor(context: Context, attrs: AttributeSet) : View(context, attrs) {
     private val paint = Paint()
     private var path = Path()
     private val drawings = mutableListOf<Drawing>()
```
저는 화면 회전은 생각 못했었습니다. 🫨 
화면 회전은 생각해 둔 대응법이 혹시 있으셨나요??

Junyoung-WON replied:
Drawing 목록을 저장하는 Drawings 와 같은 객체를 간단하게는 `object`로 만들거나, 어플리케이션에서 Drawings 객체를 생성, 관리하도록 해도 좋을 것 같아요!
그렇게 하면 구성 변경이 일어나도 이전에 그렸던 그림들을 Drawings로부터 가져올 수 있으니 비교적 간단하게 대응할 수 있지 않을까 합니다.

aprilgom commented on app/src/main/java/woowacourse/paint/ui/DrawingBoardView.kt:
```diff
@@ -9,53 +9,121 @@ import android.util.AttributeSet
     @SuppressLint("ClickableViewAccessibility")
     override fun onTouchEvent(event: MotionEvent): Boolean {
         when (event.action) {
             MotionEvent.ACTION_DOWN -> {
```
점 하나 된다 드디어!! 😆

aprilgom reviewed (changes requested):
호두두두두 안녕하세요
리뷰가 좀 늦었습니다
태블릿도, 다크모드도 잘 되네요.
구현을 잘 해주셨기에 코멘트 몇 가지만 확인해주시면 될 것 같습니다.

Junyoung-WON committed:
feature: (3단계) transfer mode 변경 및 지우개 블러 효과 적용

- PorterDuff.Mode 가 DST_OUT 이어도 지우개 효과가 적용됨
- paint의 mastFilter를 BlurMaskFilter로 설정하여 블러 효과 적용함

Junyoung-WON committed:
feature: (3단계) 구성 변경에 대응하기 위해 Drawings 싱글톤 객체 생성

- Drawing들을 저장하고 추가할 수 있는 전역 싱글톤 객체

Junyoung-WON committed:
refactor: (3단계) 그리기 모드에 따른 각 그림들이 스스로 그리기 동작을 구성하도록 변경

- paint 에 대한 설정도 그림 스스로가 변경하도록 개선

Junyoung-WON commented:
차차차람! 😄
좋은 레퍼런스와 함께 재미있는 고민, 공부거리를 던져주셔서 감사드립니다!
사실 지우개를 구현했던 코드의 내부 동작이 궁금하여 [얕은 다이브](https://github.com/woowacourse/android-paint/pull/100#discussion_r1789734241)를 해보았었는데,
**하드웨어 가속 설정을 하는 부분**에 대해서도 잘 몰랐던 것 같아요!
이번 기회에 [공부](https://github.com/woowacourse/android-paint/pull/111#discussion_r1793134819)해보니, `invalidate()`의 내부 동작과 안드로이드 시스템이 View 계층 구조를 어떻게 렌더링하는지에 대해서 자세히 알게된 것 같네요!

이렇게 공부해본게 너무 오랜만이라, 재미도 있고 시간가는줄 몰라서 리뷰 요청이 늦어졌네요...! 마치 레벨 1, 2로 돌아간 느낌... 😌
코드도 객체 지향적인 관점에서 조금 더 개선해보았습니다! 각 그림들이 스스로 자신의 책임대로 그릴 수 있도록요!
아래 개선사항들과 커밋들을 정리해보았습니다. 리뷰에 도움이 되었으면 좋겠습니다!

### 주요 개선 사항
- 지우개의 `Paint` `transfer mode` 변경 및 블러효과 적용 : [9c0fce8](https://github.com/woowacourse/android-paint/pull/111/commits/9c0fce854002f99141d3bcaaa691b0fb63f3ccea)
- 구성 변경 대응을 위한 `Drawings` 객체 생성 : [384a392](https://github.com/woowacourse/android-paint/pull/111/commits/384a392eef887dcb526181abeb20bf40755ddf67)
- 각 그림들이 스스로 `Paint`를 설정하고 그리는 동작을 구현하도록 변경 : [8e6f6cf](https://github.com/woowacourse/android-paint/pull/111/commits/8e6f6cf592362a3a48347740f7bdfc8f7ae055ba)

### 여전히 아쉬운 점
- 구성 변경에 대응해보려 했으나, 화면 회전 시 `Canvas`의 크기가 달라지면서, 기존 그림들이 위치해있던 좌표가 변경된 화면에서는 엉뚱한 곳을 가리키게 됩니다. 그래서 아래와 같이 그림들이 화면 밖으로 사라지게 됩니다.
  - 태블릿 가로 모드에서 그린 그림
    ![image](https://github.com/user-attachments/assets/b5ac6f9e-669d-4b25-81e2-e77260aeb7bd)
  - 세로 모드로 변경 시 오른쪽에 있던 그림들이 밖으로 사라짐
    ![image](https://github.com/user-attachments/assets/bee84020-e094-47e9-b2df-c9ea34776ba1)
- 여전히 UX 친화적이지 않습니다! 😂

aprilgom commented:
화면 회전 대응의 경우, path들의 좌표를 기기를 회전한 방향으로 회전시키면 될 것 같아요. 
아니면 캔버스 확대, 축소 기능을 넣어서 회전을 시켜서 그림이 가려지더라도 
그렸던 그림을 모두 볼 수 있게 하는 방법도 있을 것 같고..

aprilgom commented on app/src/main/java/woowacourse/paint/model/Circle.kt:
```diff
@@ -0,0 +1,37 @@
+
+class Circle(path: Path, paint: Paint) : Drawing(path, paint) {
+    init {
+        setPaintStyle()
```
setPaintStyle은 Drawing init할 때 실행하는 건 어떤가요

aprilgom commented on app/src/main/java/woowacourse/paint/model/Drawing.kt:
```diff
@@ -2,8 +2,36 @@ package woowacourse.paint.model
+import woowacourse.paint.model.DrawingMode.ERASE
 
-data class Drawing(
+sealed class Drawing(
```
Drawing을 상속받고 draw를 구현하는 방식으로 만드셨네요!

aprilgom reviewed (approved):
호두 고생 많으셨습니다!
파고들어 공부하는 모습 멋있었습니다. 많이 배워가요 👍 
Compose 미션도 열심히 해봅시다!!

aprilgom commented:
## 1. Path를 여러 개 그리는 문제,
path 위에 path를 그리고 그걸 블렌딩 옵션으로 가려버리는 방식이 맞는가에 대한 의문이 남아있으신 것 같습니다.
일단 이 방식도 폰에 선을 몇 만개 그릴 사람이 아닌 이상 괜찮지 않나..? 라는 생각입니다. 
저는 View에 Bitmap을 하나 두고 거기에 Path를 그리는 걸로 해결을 봤었는데요,
이러면 Path를 여러 개 그릴 필요는 없지만, 실질적으로 성능 향상이 이루어지는지는 테스트를 못해봤습니다. 
선을 몇 개 안그린다면 오히려 더 느릴지도 모르겠네요.
제 코드는 쫌 지저분하니, 케이엠의 코드를 한 번 참고해보는게 좋을 것 같아요~

## 2. 지우개에 대한 의문
이왕 비트맵에 이미지를 그리는 겸, 이쁜 브러쉬를 써보는 것도 생각을 해봤습니다. 
포토샵 같은 툴을 보면 브러쉬가 상당히 다양하거든요. 
그런 브러쉬를 그리려면 그냥 선을 긋는게 아닌 Path를 따라서 브러쉬 이미지를 여러 개 뿌려야 할 것 같습니다.
![unnamed](https://github.com/user-attachments/assets/87e5cf81-4941-4ed9-aafa-49fae19b0411)
(모양이 겁나 다양하죠)

그런데..! 문제는 지우개도 이런 겁나 다양한 모양을 지원해야 한다는 겁니다 지엔장~~🤮
지우개도 펜과 동일하게 비트맵을 뿌리고, 우리가 지금까지 지우개를 썼던 것처럼 blend mode를 적용시키면 될 것 같네요.
지우개와 PorterDuff에 대한 질문은 이런 맥락에서 나왔습니다. 
저런 다양한 모양의 비트맵 지우개의 구현이 PorterDuff만으로 될까?
결론만 짧게 얘기하면, 제 생각엔 안될 것 같습니다.
PorterDuff의 DST_OUT과 CLEAR로는 안되고, [아마 Subtract가 필요할 것 같아요.](https://en.wikipedia.org/wiki/Blend_modes#Subtract)
